### PR TITLE
Workflows: Adding new workflow (with test cases) - Verify_Register_SVD.

### PR DESCRIPTION
**Pull request by sihyung-maxim**
**Overview:**
The new Verify_Register_SVD workflow keeps track of any register (e.g. i2c_regs.h) or peripheral SVD (e.g. i2c_reva_me17.svd) file changes in a PR, and makes sure all register files are synced to their appropriate peripheral SVD file before allowing you to merge. This workflow checks the parts' register files, not the private RevX register files like i2c_reva_regs.h or tmr_revb_regs.h.

The test cases are commented below. You may have to expand the log history if some of the comments were hidden away.

**Clarifying naming convention for this workflow** 
- Both "**peripheral SVD files**" and "**SVD files**" refer to the individual peripheral SVD files used to generate a register file - e.g. i2c_reva_me14.svd, gcr_es17.svd, and tmr_revb.svd. "**SVD files**" does not include the parts' SVD file - e.g. max32665.svd. 
- A "**_generic_ peripheral SVD file**" or "**_generic_ SVD file**" describes a single peripheral SVD file used to generate register files for multiple parts - e.g. icc_reva.svd -> icc_regs.h for ME12, ME14, ME17, AI85, ... etc. The SVD file name does not include the part name - e.g. icc_reva.svd, tmr_revb.svd, i2c_reva.svd.
- A "**_part-specific_ peripheral SVD file**" or "**_part-specific_ SVD file**" refers to a unique peripheral SVD file used to generate a register file for (generally) only One part - e.g. spi_reva_me15.svd -> spi_regs.h for only ME15. The SVD file name includes the part name - e.g. gpio_reva_me12.svd, dma_reva_me10.svd, dma_reva_me11.svd.  There can be exceptions where there are multiple parts with register files pointing to the same "**_part-specific_ SVD file**". For example, both the ME15's and ME16's gpio_regs.h file depends on gpio_reva_me15.svd.

**Purpose:** 
The purpose of this new workflow is to make sure the register files and peripheral SVD files are always synced and up to date for our users, the user guides, and the debugging watch windows (e.g. max32655.svd for the register view window in debuggers).

Our microcontroller user guides reference the SVD files in our MSDK. When there's a bug fix made to a register file in the MSDK, the corresponding peripheral SVD file is rarely updated with the register file - causing the MSDK and user guide to show conflicting information. To minimize any headache and confusion, the Verify_Register_SVD workflow verifies both register and SVD files are synced, so any register changes/fixes are quickly brought into the next version of the user guide.

**Function:**
Once a runner picks up this job and sets up its local workspace, the job scans for any register or peripheral SVD file changes. If there are no changes, then the job quickly runs to completion and the PR passes this workflow check. However, if there are any detected changes, the job will compare all the register-to-SVD file sets. The check will only pass if all the register files are synced to the peripheral SVD file, and will fail otherwise.

Luckily, the workflow doesn't leave you hanging in the dark if the PR's register/SVD files aren't synced. The job's log output shows all the relevant files that need to be updated before the PR can pass this check. 

For example, if you change this field in *Libraries/CMSIS/Device/Maxim/MAX32655/Include/i2c_regs.h* from: 
`#define MXC_F_I2C_CTRL_EN_POS                          0 /**< CTRL_EN Position */`
to:
`#define MXC_F_I2C_CTRL_EN_POS                          1 /**< CTRL_EN Position */`
Then, the job's log will tell you the *Libraries/PeriphDrivers/Source/I2C/i2c_reva_me17.svd* file needs to be updated with this same field change to sync up the two files. The reverse also works if you only update *i2c_reva_me17.svd* and not *i2c_regs.h*.

**Notes:**
The test case status descriptions ([FAILED] and [PASSED]) in the comments below do not reflect the log output of the job.

I recreated this PR, so the reviewers have an easier time looking at all the test cases. The original PR had too many commit lines which caused some of my important comments to be hidden. For documentation purposes, here is the original PR and deleted branch of this workflow:

https://github.com/Analog-Devices-MSDK/msdk/pull/236
https://github.com/Analog-Devices-MSDK/msdk/tree/workflows/verify_register_svd

### CODE REVIEW

**Comment by sihyung-maxim** [comment]
Test 1: Deleted an important peripheral SVD file required by the SVD scripts to run properly. "Important peripheral SVD files" include all peripheral SVD files listed in chip_periph.txt.

[FAILED] Missing peripheral SVD file.

https://github.com/Analog-Devices-MSDK/msdk/actions/runs/3515628407/jobs/5891141202

**Comment by sihyung-maxim** [comment]
Test 1: Resolve this failed check by either reverting deleted changes or adding back the file.

[PASSED] Reverted deleted changes.

https://github.com/Analog-Devices-MSDK/msdk/actions/runs/3515655076/jobs/5891202740

**Comment by sihyung-maxim** [comment]
Test 2: Adding a new peripheral SVD file that's not used by the SVD scripts - meaning not listed in chip_periph.txt.

[PASSED] No register file depends on this new peripheral SVD file nor do the SVD scripts require this new file.

https://github.com/Analog-Devices-MSDK/msdk/actions/runs/3515745140/jobs/5891398786

Note - A new peripheral SVD file should be added with the intent of generating a register file for a part, so adding a new SVD file that won't be used to generate any register file should generally be avoided. There could be a scenario where a new peripheral library was created called NEW - Libraries/PeriphDrivers/NEW. This library will require a private register file - _new_reva_regs.h_, and a generic SVD file that was used to generate the private register file - _new_reva.svd_. However, the generic SVD file might not be used to generate any register file for all supported parts, so part-specific SVD files are included - _new_reva_meXX.svd_ and _new_reva_meYY.svd_. The test passes because this change is benign to the register files, and the generic _new_reva.svd_ file is still required to generate the private register file _new_reva_regs.h_.

**Comment by sihyung-maxim** [comment]
Test 3: Adding a new peripheral SVD file that is used by the SVD scripts - meaning the new SVD file is listed in chip_periph.txt. Before this test, the ME20's chip_periph.txt was updated to include this new SVD file. This is done internally, and only developers should be adding new peripheral SVD files.

ME20: qdec_reva_me20.svd

[FAILED] Missing corresponding register file - qdec_regs.h for ME20.

https://github.com/Analog-Devices-MSDK/msdk/actions/runs/3516391319/jobs/5892891136

**Comment by sihyung-maxim** [comment]
Test 4: Added a deprecated file to show the workflow can detect them.

[FAILED] AI85 bbfc_regs.h is a deprecated file.

https://github.com/Analog-Devices-MSDK/msdk/actions/runs/3516447923/jobs/5893019898

**Comment by sihyung-maxim** [comment]
Test 4: Deleting deprecated file.

[FAILED] Unsupported register file.

https://github.com/Analog-Devices-MSDK/msdk/actions/runs/3516710573/jobs/5893618866

**Comment by sihyung-maxim** [comment]
Test 4: Revert all changes made to the deprecated register file.

[PASSED[ Reverted changes.

https://github.com/Analog-Devices-MSDK/msdk/actions/runs/3516742598/jobs/5893689832

**Comment by sihyung-maxim** [comment]
Test 5: This shows what happens when you only change a register file (in this case - ME15's gpio_regs.h) but not its "corresponding peripheral SVD file" (gpio_reva_me15.svd). Not only that, the workflow will find register files in the other parts that are dependent on said "corresponding peripheral SVD file", and make sure all affected files are tracked and synced.

This is also a special case where two parts (ME15 and ME16) are dependent on a part-specific SVD file.

ME15: gpio_reva_me15.svd -> gpio_regs.h
ME16: gpio_reva_me15.svd -> gpio_regs.h

[FAILED] gpio_reva_me15.svd and gpio_regs.h doesn't match for both ME15 and ME16

https://github.com/Analog-Devices-MSDK/msdk/actions/runs/3516824624/jobs/5893864902

**Comment by sihyung-maxim** [comment]
Test 5: Synced up all the affected files tracked by the workflow - ME15 gpio_regsh, ME16 gpio_regs.h, and gpio_reva_me15.svd.

[PASSED] ME15 and ME16 GPIO register and SVD files are synced

https://github.com/Analog-Devices-MSDK/msdk/actions/runs/3516935191/jobs/5894116409

**Comment by sihyung-maxim** [comment]
Test 6: Changed a single register file with a corresponding **part-specific** peripheral SVD file. No other register file should be dependent on the **part-specific** SVD file for this test case. One register file to One SVD file.

ME13: dma_reva_me13.svd -> dma_regs.h

[FAILED] Register and SVD file doesn't match.

https://github.com/Analog-Devices-MSDK/msdk/actions/runs/3516977411/jobs/5894205036

**Comment by sihyung-maxim** [comment]
Test 6: Synced the one-to-one register and **part-specific** SVD file.

[PASSED] dma_reva_me13.svd and dma_regs.h are synced for the ME13.

https://github.com/Analog-Devices-MSDK/msdk/actions/runs/3517012849/jobs/5894283356

**Comment by sihyung-maxim** [comment]
Test 7: Changed a single register file with a corresponding **generic** peripheral SVD file - meaning a single SVD file is used to generate a register file for multiple parts.

ME14: sema_regs.h -> sema_reva1.svd
ME13: sema_reva1.svd -> sema_regs.h
ME10: sema_reva1.svd -> sema_regs.h

[FAILED] Register and SVD files aren't synced for ME10, ME13, and ME14.

https://github.com/Analog-Devices-MSDK/msdk/actions/runs/3517034688/jobs/5894333962

**Comment by sihyung-maxim** [comment]
Test 7: Synced up all affected files.

[PASSED] SEMA RevA1 SVD file and all dependent register files are synced for ME10, ME13, and ME14.

https://github.com/Analog-Devices-MSDK/msdk/actions/runs/3517088241/jobs/5894450363

**Comment by sihyung-maxim** [comment]
Test 7b: Continuing on from the original test 7, this time only the corresponding **generic** SVD file was synced to the ME14's changed sema_regs.h, and not all the register files dependent on said **generic** SVD file. This test case will show what happens when only some of the register and SVD files were synced, but not all.

Only ME14 sema_regs.h and sema_reva1.svd were synced. The ME10 and ME13 sema_regs.h is out of sync.

Changed: ME14's sema_regs.h; sema_reva1.svd.
ME14: sema_regs.h -> sema_reva1.svd
ME13: sema_reva1.svd -> sema_regs.h
ME10: sema_reva1.svd -> sema_regs.h

[FAILED] ME10 and ME13 sema_regs.h still needs to sync up to sema_reva1.svd.

https://github.com/Analog-Devices-MSDK/msdk/actions/runs/3517159667/jobs/5894609578

**Comment by sihyung-maxim** [comment]
Test 8: Changed only a **part-specific** peripheral SVD file with only one register file dependency. 

Changed: wut_reva_me18.svd
ME18: wut_reva_me18.svd -> wut_regs.h

[FAILED] Register and SVD files aren't synced for ME18.

https://github.com/Analog-Devices-MSDK/msdk/actions/runs/3517267340/jobs/5894835001

**Comment by sihyung-maxim** [comment]
Test 8: Synced up register file to **part-specific** peripheral SVD file.

[PASSED] wut_regs.h and wut_reva_me18.svd are synced for ME18.

https://github.com/Analog-Devices-MSDK/msdk/actions/runs/3517281594/jobs/5894866135

**Comment by sihyung-maxim** [comment]
Test 9: Changed a single **generic** peripheral SVD file.

Changed: afe_hart.svd
ME16: afe_hart.svd -> afe_hart_regs.h
ME20: afe_hart.svd -> afe_hart_regs.h

[FAILED] afe_hart.svd not synced with any corresponding register file.

https://github.com/Analog-Devices-MSDK/msdk/actions/runs/3517547904/jobs/5895446429

**Comment by sihyung-maxim** [comment]
Test 9: Synced up all dependent register files to **generic** peripheral SVD file.

[PASSED] afe_hart_regs.h and afe_hart.svd is synced for ME16 and ME20.

https://github.com/Analog-Devices-MSDK/msdk/actions/runs/3517563265/jobs/5895476845

**Review by Jacob-Scheiffler — APPROVED** [review]
Since, I don't know anything about writing bash scripts, I'll leave the review of the actual code to Lorne and Kevin. As far as test cases go, I think you're good. I can't think of any you missed.

File: .github/workflows/Verify_Register_SVD.yml [hunk 1]
```diff
@@ -0,0 +1,573 @@
...
+          fetch-depth: 0
+          token: ${{ secrets.WORKFLOW_TOKEN }}
+          path: ${{ env.MSDK-INTERNAL_DIR }}
+          
+      # Use these next steps to verify if register files were updated.
+      # Easier to read in Actions' console window if separated by steps.
+      - name: Checking if register files were updated.
+        run: |
+          cd ${{ env.MSDK_DIR }}
+          CHANGED_FILES=$(git diff --ignore-submodules --name-only remotes/origin/main './Libraries/CMSIS/Device/Maxim/*/Include/*_regs.h' './Libraries/PeriphDrivers/Source/*/*.svd')
```
Comment (kevin-gillespie) [inline .github/workflows/Verify_Register_SVD.yml:73]: Will this catch changes to Libraries/PeriphDrivers/Source/SYS/SVD/*.svd?

**Review by kevin-gillespie — COMMENTED** [review]
Are we checking the SVD schema?

**Comment by sihyung-maxim** [comment]
A simple fix was made to correctly parse a file name to isolate its peripheral name. (gcr_me15.svd -> gcr).

**Comment by sihyung-maxim** [comment]
Test 10: Changing a SYS SVD File in Libraries/PeriphDrivers/Source/SYS/SVD/*.svd.

Changed: gcr_me15.svd

[FAILED] gcr_regs.h for ME15 and ME16 are not synced with gcr_me15.svd.

https://github.com/Analog-Devices-MSDK/msdk/actions/runs/3568411260/jobs/5997251665

**Comment by sihyung-maxim** [comment]
Test 10:

[PASSED] The ME15 and ME16 gcr_regs.h are synced with gcr_me15.svd.

https://github.com/Analog-Devices-MSDK/msdk/actions/runs/3568493299/jobs/5997421759

File: .github/workflows/Verify_Register_SVD.yml (hunk 1 above)
Comment (sihyung-maxim) [reply .github/workflows/Verify_Register_SVD.yml:73]: Yes, this catches any changes in Libraries/PeriphDrivers/Source/SYS/SVD/*.svd. The asterisk symbol represented as a directory (in './Libraries/PeriphDrivers/Source/*/*.svd') includes all directories and their subdirectories, so the git diff can find any SVD file change in Libraries/PeriphDrivers/Source, no matter the depth which the SVD file is located.

Test 10 below demonstrates what happens when you change a SYS SVD file.

**Comment by lorne-maxim** [comment]
Two questions related to the dictionaries for part numbers and die types:

1. If a new part is added to the SDK, but the dictionaries are not updated, will the workflow pass?  Generating a failure would be preferred so there is a clear reminder that the dictionaries need to be updated.
2. How do the dictionaries work if one die type is used by multiple part numbers or vice versa?  For example, don't some of the MAX78000 peripherals use the ME17 files in addition to AI85 files?  Aren't some ME15 files used by the MAX32670 and MAX32675?

**Comment by sihyung-maxim** [comment]
Test 11: Changing/Adding a register file for a "new" part that hasn't been added to the workflow's dictionary.

Added: MAX78010 adc_regs.h

[FAILED] Unknown part detected.

https://github.com/Analog-Devices-MSDK/msdk/actions/runs/3587509371/jobs/6037935872

**Comment by kevin-gillespie** [comment]
Are we checking the SVD schema?

**Comment by sihyung-maxim** [comment]
> Two questions related to the dictionaries for part numbers and die types:
> 
> 1. If a new part is added to the SDK, but the dictionaries are not updated, will the workflow pass?  Generating a failure would be preferred so there is a clear reminder that the dictionaries need to be updated.
> 2. How do the dictionaries work if one die type is used by multiple part numbers or vice versa?  For example, don't some of the MAX78000 peripherals use the ME17 files in addition to AI85 files?  Aren't some ME15 files used by the MAX32670 and MAX32675?

1. Good point. The workflow will fail in that case, but I've made some changes to the log output to explicitly state what's wrong (See Test 11).
2. The MSDK CMSIS directories use the chip name while the part-specific peripheral SVD files and the _Device_ directories for the SVD scripts both use the die name. The dictionaries exist to easily switch between the chip name and die name when traversing the workspace, and they aren't directly used to find the parts' SVD file dependencies. Each part has a chip_periph.txt file, located with the SVD scripts, which lists the paths to all the parts' required peripheral SVD files.

**Comment by sihyung-maxim** [comment]
> Are we checking the SVD schema?

Are you referring to the part SVD files (e.g. max32665.svd) and the svd_schema.xsd file?

**Comment by kevin-gillespie** [comment]
> > Are we checking the SVD schema?
> 
> Are you referring to the part SVD files (e.g. max32665.svd) and the svd_schema.xsd file?

Yes, that's it

**Comment by sihyung-maxim** [comment]
> > > Are we checking the SVD schema?
> > 
> > 
> > Are you referring to the part SVD files (e.g. max32665.svd) and the svd_schema.xsd file?
> 
> Yes, that's it

No, this workflow doesn't track those files. The issue is that if a generic, peripheral SVD file gets updated, then all the parts linked to that SVD file will need to have their <part>.svd file updated. That might be asking for too much in one check. Once this workflow is merged in, I was planning to create another job that would check for the SVD schema, and let a runner automatically update the part SVD file once this current workflow is triggered and all the checks pass.

**Comment by kevin-gillespie** [comment]
> > > > Are we checking the SVD schema?
> > > 
> > > 
> > > Are you referring to the part SVD files (e.g. max32665.svd) and the svd_schema.xsd file?
> > 
> > 
> > Yes, that's it
> 
> No, this workflow doesn't track those files. The issue is that if a generic, peripheral SVD file gets updated, then all the parts linked to that SVD file will need to have their <part_number>.svd file updated. That might be asking for too much in one check. Once this workflow is merged in, I was planning to create another job that would check for the SVD schema, and let a runner automatically update the part SVD file if this current workflow is triggered and all the checks pass.

Ok sounds good.

**Review by kevin-gillespie — APPROVED** [review]

**Pull request merged**